### ISSUE TITLE
Visual C++ language version

### ISSUE THREAD

**Issue by Timmmm** [mesonbuild/meson#1973]
As far as I can tell, `cpp_std` does nothing with MSVC. It's also not clear what it should do if you specify `gnu...` versions, e.g. `gnu++11`. I suggest it should treat those as if they were `c++11`, otherwise there's no way to get `gnu++14` with GCC, and `c++14` with MSVC.

These are the [available options](https://blogs.msdn.microsoft.com/vcblog/2016/06/07/standards-version-switches-in-the-compiler/). As of now it is just `/std:c++14` and `/std:c++latest`. Probably not worth supporting `c++latest`.

**Comment by nirbheek** [comment]
Yes, this is a known missing feature. We can do two things here which are reasonable:

1. Maintain a mapping between VC++ version and the stds they support

2. If the requested std is older than c++14, validate against the mapping and error out if the compiler version is too old. If it's new enough, no new args will be added.

3. If the requested std is c++14 (or newer in the future) and the compiler version is new enough, add -std:c++14. If the compiler version is too old for the option but new enough for c++14, do nothing. If the compiler version is too old for c++14, error out.

However, it's not clear what the advantage is over just ignoring the `cpp_std` value right now. It looks like `-std:c++14` doesn't actually do anything yet. It just reflects the current state of the compiler. In fact, [it reflects the current state rather poorly](https://msdn.microsoft.com/en-us/library/mt490614.aspx). Even in [VS 2017](https://docs.microsoft.com/en-gb/cpp/build/reference/std-specify-language-standard-version).

I think we should wait and see if Microsoft actually ever releases a C++ compiler that can be run in standards-compliant mode.

**Comment by polarina** [comment]
I implemented support for this in https://github.com/mesonbuild/meson/pull/2836

**Comment by JamBalaya56562** [comment]
This is implemented — it was added in
[PR #2836](https://github.com/mesonbuild/meson/pull/2836), taking the approach
@nirbheek described above.

Today, in
[`mesonbuild/compilers/cpp.py`](https://github.com/mesonbuild/meson/blob/master/mesonbuild/compilers/cpp.py),
`cpp_std` maps to the MSVC switches via a `VC_VERSION_MAP` in `get_option_std_args`:

- `c++14` / `vc++14` → `/std:c++14`, `c++17` / `vc++17` → `/std:c++17`,
  `c++20` / `vc++20` → `/std:c++20`, and `c++latest` / `vc++latest` →
  `/std:c++latest`.
- The plain `c++NN` values also add `/permissive-` (standards-conforming mode);
  the `vc++NN` variants select the same `/std:` switch *without* `/permissive-`,
  so both MSVC modes are reachable.
- **`c++11`** is handled the way you suggested rather than being ignored: MSVC
  has no `/std:c++11` switch, so Meson maps it to `/std:c++14` (via
  `CPP11AsCPP14Mixin`) and emits a one-time warning, instead of silently adding
  nothing.

The accepted `cpp_std` values are compiler-specific, so an MSVC build selects
from the `c++NN` / `vc++NN` set above while GCC/Clang keep their
`c++NN` / `gnu++NN` values — which gives the per-toolchain standard selection
this issue was after.

From the above this looks resolved and can be closed as implemented. If a
specific value isn't producing the expected `/std:` flag for you, please comment
and it can be revisited.
